### DISCUSSION TITLE
Logos in media folder

### DISCUSSION THREAD

**Discussion by hcwinsemius** [fabiocaccamo/django-admin-interface#383] [Q&A] [answered]
Hello all,

I want to serve media files over a S3 bucket, but don't want the media to be entirely public. But I do need the logos public! They are stored in the media folder. Does anybody know how to achieve this?

Love the work!

**Comment by fabiocaccamo** [comment]
@hcwinsemius this could be achieved by supporting a new optional setting `ADMIN_INTERFACE_LOGO_STORAGE` that defaults to `'django.core.files.storage.FileSystemStorage'`.

**Reply by hcwinsemius** [reply]
I implemented that option by pointing any media in my own app to an additional storage configured in `settings.py` in the `STORAGES` option. Works perfectly that way now but indeed it would be very neat to have that as optional parameter under this package.

**Reply by fabiocaccamo** [reply]
> I implemented that option by pointing any media in my own app to an additional storage configured in `settings.py` in the `STORAGES` option. Works perfectly that way now but indeed it would be very neat to have that as optional parameter under this package.

That's a solution indeed, but it affects all your media files.

**Comment by cor14095** [comment, marked as answer]
Hi, I stumble through this issue and got around it kinda simple. But It was hard for me to figure out the pieces so I'll leave this here and look to open a PR to enhance documentarion. 

First of all in my `settings.py` I have something like this:
```
# AWS Configuration
    AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID")
    AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
    AWS_STORAGE_BUCKET_NAME = os.getenv("AWS_STORAGE_BUCKET_NAME", "example-bucket")
    AWS_S3_REGION_NAME = os.getenv("AWS_S3_REGION_NAME", "us-east-1")

# Default file storage (This is kinda the important part)
    DEFAULT_FILE_STORAGE = os.getenv("DEFAULT_FILE_STORAGE", "django.core.files.storage.FileSystemStorage")
```

The important part of this is configuring the `DEFAULT_FILE_STORAGE` to the s3Boto3Storage in your `.env`:
```
AWS_ACCESS_KEY_ID=...
AWS_SECRET_ACCESS_KEY=...
AWS_STORAGE_BUCKET_NAME=my-bucket
AWS_S3_REGION_NAME=us-east-1

# This is the important config
DEFAULT_FILE_STORAGE=storages.backends.s3boto3.S3Boto3Storage
```

With these env variables set I later added an S3 config in the settings:
```
# S3 storage configuration
if AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY:
    # AWS_DEFAULT_ACL = "public-read"  # I don't use ACL so I don't use it, but you can have it set here.

    # I generate an S3 custom domain here because I manage multiple projects in a single bucket, but you can generate a static url here.
    # I just find this to be more organized.
    AWS_S3_CUSTOM_DOMAIN = (
        f"{AWS_STORAGE_BUCKET_NAME}.s3.{AWS_S3_REGION_NAME}.amazonaws.com"
    )
    AWS_LOCATION = "media"  # You can change this base on your desire media directory.

    # Media URL will point to S3
    MEDIA_URL = f"https://{AWS_S3_CUSTOM_DOMAIN}/{AWS_LOCATION}/"
    # I don't set MEDIA_ROOT here since files are provisioned by an s3 url.

    # S3 upload settings
    AWS_S3_FILE_OVERWRITE = False  # Don't overwrite files with the same name (optional)
    AWS_QUERYSTRING_AUTH = False  # Don't add query parameters to URLs (optional)
else:
    # If the S3 config fails then I fallback to defaults. Not needed I think but I do it. 
    MEDIA_URL = "/media/"
    MEDIA_ROOT = os.path.join(BASE_DIR, "media")
```

By doing this configurations I was able to change the default behavior of the Logo file storage and handle the S3 implementation. This is more of an S3/Django default configurations. But I think it's worth to have it documented somewhere.

**Reply by merwok** [reply]
That’s one of the common ways to use different storages for local environment vs server environments.
But it doesn’t address the need discussed here: different storages for different models, in the same env!

**Reply by cor14095** [reply]
This is a quick GP solution but it would need to have a custom storage class like:
```
# myapp/storage_backends.py
from django.conf import settings
from django.core.files.storage import FileSystemStorage
from storages.backends.s3boto3 import S3Boto3Storage
from django.core.files.storage import Storage

class S3OrLocalStorage(Storage):
    """
    Custom storage that delegates to S3Boto3Storage if USE_S3_STORAGE is True,
    otherwise to FileSystemStorage.
    """
    def __init__(self, *args, **kwargs):
        if getattr(settings, "USE_S3_STORAGE", False):
            self.storage = S3Boto3Storage()
        else:
            self.storage = FileSystemStorage()
    
    def _open(self, name, mode="rb"):
        return self.storage._open(name, mode)
    
    def _save(self, name, content):
        return self.storage._save(name, content)
    
    def exists(self, name):
        return self.storage.exists(name)
    
    def url(self, name):
        return self.storage.url(name)
    
    def delete(self, name):
        return self.storage.delete(name)
    
    def size(self, name):
        return self.storage.size(name)
    
    def accessed_time(self, name):
        return self.storage.accessed_time(name)
    
    def created_time(self, name):
        return self.storage.created_time(name)
    
    def modified_time(self, name):
        return self.storage.modified_time(name)
```

Then change the logo for something like:
```
from storage_backends import S3OrLocalStorage

[...]

logo = models.FileField(
        upload_to="admin-interface/logo/",
        storage=S3OrLocalStorage(),
        blank=True,
        validators=[
            FileExtensionValidator(
                allowed_extensions=["gif", "jpg", "jpeg", "png", "svg"]
            )
        ],
        help_text=_("Leave blank to use the default Django logo"),
        verbose_name=_("logo"),
    )
```

This would only need a settings variable like:
`USE_S3_STORAGE = True  # or False for local filesystem storage` 

Would this solution work for everyone? It's fairly simple to implement. The big Issue I see here is that it would make boto3 a requirement even if you don't use it.

**Reply by merwok** [reply]
No, I don’t think a reusable app can depend on a specific storage class. This is not the way.

**Reply by cor14095** [reply]
Is there any other way for this? I'm not the most knowledgeable on django, but from the docs and google searching, I don't seem to find any other plausible solutions. 

Do you have any idea what it could be made? I'll try to ask around and do some research. I'll be back if I find anything that can allow multiple storages without making boto3 a requirement.

**Reply by merwok** [reply]
See the first reply by the library author

**Reply by fabiocaccamo** [reply]
I created an issue: https://github.com/fabiocaccamo/django-admin-interface/issues/428
